File: algorithms/449. Serialize and Deserialize BST/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # @StefanPochmann
        def encode(cur):
            if cur:
                nums.append(str(cur.val))
                encode(cur.left)
                encode(cur.right)
        nums = []
        encode(root)
        return ' '.join(nums)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        preorder = map(int, data.split())
        inorder = sorted(preorder)
        preorder.reverse()
        inorder.reverse()
        def build(stop):
            if inorder and inorder[-1] != stop:
                root = TreeNode(preorder.pop())
                root.left = build(root.val)
                inorder.pop()
                root.right = build(stop)
                return root
        return build(None)

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

# 62 / 62 test cases passed.
# Status: Accepted
# Runtime: 158 ms
# 48.00%

File: algorithms/449. Serialize and Deserialize BST/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        // based on @code9yitati
        string out;
        char buffer[1024];
        dfsEncode(root, out, buffer);
        return out;
    }

    void dfsEncode(TreeNode * root, string & out, char * buffer) {
        if ( root ) {
            memcpy(buffer, & (root->val), sizeof(int));
            out.insert(out.end(), buffer, buffer + sizeof(int));
            dfsEncode(root->left, out, buffer);
            dfsEncode(root->right, out, buffer);
        }
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int pos = 0;
        return dfsDecode(data, pos, INT_MIN, INT_MAX);
    }

    TreeNode * dfsDecode(const string & in, int & pos, int minVal, int maxVal) {
        if ( pos >= in.size() ) {
            return NULL;
        }
        int val;
        memcpy(& val, & in[pos], sizeof(int));
        if ( val < minVal || val > maxVal ) {
            return NULL;
        }
        TreeNode * root = new TreeNode(val);
        pos += sizeof(int);
        root->left = dfsDecode(in, pos, minVal, val);
        root->right = dfsDecode(in, pos, val, maxVal);
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

// 62 / 62 test cases passed.
// Status: Accepted
// Runtime: 19 ms
// beats 97.10 %
